core/src/main/java/com/linecorp/armeria/client/Endpoint.java

```diff
@@ -283,21 +409,36 @@ public boolean equals(Object obj) {
         final Endpoint that = (Endpoint) obj;
         if (isGroup()) {
             if (that.isGroup()) {
-                return authority().equals(that.authority());
+                return groupName().equals(that.groupName());
             } else {
                 return false;
             }
         } else {
             if (that.isGroup()) {
                 return false;
             } else {
-                return authority().equals(that.authority()) && weight() == that.weight();
+                return host().equals(that.host()) &&
+                       Objects.equals(ipAddr, that.ipAddr) &&
+                       port == that.port;
             }
         }
     }
 
     @Override
     public int hashCode() {
-        return authority().hashCode();
+        return (authority().hashCode() * 31 + Objects.hashCode(ipAddr)) * 31 + port;
+    }
+
+    @Override
+    public String toString() {
+        final ToStringHelper helper = MoreObjects.toStringHelper(this).omitNullValues();
+        helper.addValue(authority());
+        if (!isGroup()) {
+            if (hostIpAddrType == null) {
```

Comment:
I'm not sure but `hostIpAddrType != null`? Or did you intend writing non-IP address?

Reply:
if `hostIpAddrType` is not `null`, it means `host` is an IP address, and we already did `addValue()` with it, so we do not add it again.